src/main/java/javaslang/collection/Traversable.java

```diff
@@ -553,6 +540,77 @@ default T fold(T zero, BiFunction<? super T, ? super T, ? extends T> op) {
     Option<T> headOption();
 
     /**
+     * Returns the index of the first occurrence of the given element or -1 if this does not contain the given element.
+     *
+     * @param element an element
+     * @return the index of the first occurrence of the given element
+     */
+    default int indexOf(T element) {
+        return indexOf(element, 0);
+    }
+
+    /**
+     * Returns the index of the first occurrence of the given element after or at some start index
+     * or -1 if this does not contain the given element.
+     *
+     * @param element an element
+     * @param from start index
+     * @return the index of the first occurrence of the given element
+     */
+    int indexOf(T element, int from);
+
+    /**
+     * Finds first index where this sequence contains a given sequence as a slice.
+     * <p>
+     * Note: may not terminate for infinite-sized collections.
+     * @param that the sequence to test
+     * @return the first index such that the elements of this sequence starting at this index match
+     * the elements of sequence that, or -1 of no such subsequence exists.
+     * @throws NullPointerException if {@code that} is null.
+     */
+    default int indexOfSlice(Iterable<? extends T> that) {
+        Objects.requireNonNull(that, "that is null");
+        return indexOfSlice(that, 0);
+    }
+
+    /**
+     * Finds first index after or at a start index where this sequence contains a given sequence as a slice.
+     * <p>
+     * Note: may not terminate for infinite-sized collections.
+     * @param that the sequence to test
+     * @param from the start index
+     * @return the first index >= from such that the elements of this sequence starting at this index match
+     * the elements of sequence that, or -1 of no such subsequence exists.
+     * @throws NullPointerException if {@code that} is null.
+     */
+    default int indexOfSlice(Iterable<? extends T> that, int from) {
```

Comment:
Hi @danieldietrich ,
I meant `Iterable<?>` which would indeed be a workaround for the missing `<U super T>` in Java. I think in general it's find to have type constraint on `indexOf`, `containsAll` and the like, even though it may render certain alien usage patterns cumbersome. It's just something that should be considered briefly when an API is designed to make sure we don't miss some usage scenarios.
Best, Sebastian